app/controllers/projects_controller.rb

```diff
@@ -20,14 +20,42 @@ def new
     @project = Project.new
   end
 
-  # TODO: Limit to popular/recent projects
   def index
     if params[:id]
       @user = User.find_by username: params[:id]
       @projects = @user.projects
       render :user_index
+      return
     else
-      @projects = Project.inspiring_projects_for current_user.id
+      case params[:sort]
+      when 'stars'
+        @projects = Project.joins("LEFT OUTER JOIN rating_caches ON rating_caches.cacheable_id = projects.id")
+          .order("rating_caches.avg desc")
+      when 'forks'
+        @projects = Project.joins("LEFT OUTER JOIN projects p1 on projects.id = p1.ancestry")
+          .group("projects.id")
+          .order("count(p1.ancestry) desc")
+      when 'followers'
+        @projects = Project.joins("LEFT OUTER JOIN project_followers ON project_followers.project_id = projects.id")
+          .group("projects.id")
+          .order("count(project_followers.project_id) desc")
+      when 'last updated'
+        @projects = Project.order('updated_at DESC')
+      when 'activity'
+        @projects = Project.joins("LEFT OUTER JOIN comments ON comments.polycomment_id = projects.id
+                                   AND comments.polycomment_type='project'
+                                   LEFT OUTER JOIN issues ON issues.project_id = projects.id
+                                   AND issues.status=1")
+                            .where("comments.created_at > ? OR issues.updated_at > ?", 10.days.ago, 10.days.ago)
+                            .group("projects.id")
+                            .order("count(comments.polycomment_id)+4*count(issues.project_id) desc")
+      else
+        @projects = Project.order('created_at DESC')
+      end
+    end
```

Comment:
@rohitpaulk Do these queries look okay? They take less than 1ms but I am not sure how are they going to scale.

Also, Any tips about how can I avoid writing really lengthy test for this. As far as I can think, I will have to make multiple projects and then add multiple followers, comments, forks, issues and stars to each one of them. That must take atleast 200 lines.